### ISSUE TITLE
0.10.1: Unit tests are failing on armhf, hurd-i386 and kfreebsd-amd64

### ISSUE THREAD

**Issue by onovy** [vifm/vifm#462]
Hi,

from release 0.10.1 there are unit tests failing on some Debian archs:

armhf: https://buildd.debian.org/status/fetch.php?pkg=vifm&arch=armhf&ver=0.10.1-1&stamp=1564739057&raw=0

hurd-i386: https://buildd.debian.org/status/fetch.php?pkg=vifm&arch=hurd-i386&ver=0.10.1-1&stamp=1564741012&raw=0

kfreebsd-amd64: https://buildd.debian.org/status/fetch.php?pkg=vifm&arch=kfreebsd-amd64&ver=0.10.1-1&stamp=1564742385&raw=0

Thanks for help.

**Comment by xaizek** [comment]
Hi,

Just as an overview:

> armhf

This looks like something that happened previously on one of those builders, but it didn't reproduce.

> hurd-i386

This might also be spurious, it fails under valgrind for unknown reason, I assumed it simulates things differently. Couldn't really figure out what causes it. Although, if it's stable on Hurd, that might give a chance of debugging it.

> kfreebsd-amd64

This one looks weird, FUSE isn't actually used, so it fails for some other reason. I had to disable these new FUSE tests on OS X, because it fails to start executables there for no apparent reason (works in other tests), but this is unlikely the same issue.

**Comment by xaizek** [comment]
This patch fixes that test on hurd-i386:
```diff
 tests/iop/cp.c | 14 +++++++++++---
 1 file changed, 11 insertions(+), 3 deletions(-)

diff --git a/tests/iop/cp.c b/tests/iop/cp.c
index 5ab34af7e..72e953bef 100644
--- a/tests/iop/cp.c
+++ b/tests/iop/cp.c
@@ -9,7 +9,7 @@
 #include <unistd.h> /* _Exit() lstat() */
 
 #include <signal.h> /* SIGXFSZ SIG_IGN signal() */
-#include <stdlib.h> /* EXIT_SUCCESS */
+#include <stdlib.h> /* EXIT_FAILURE EXIT_SUCCESS */
 
 #include "../../src/compat/fs_limits.h"
 #include "../../src/compat/os.h"
@@ -506,8 +506,16 @@ TEST(append_truncates_destination_files_on_error, IF(not_windows))
 		const struct rlimit rlim = { .rlim_cur = 25, .rlim_max = 25 };
 
 		(void)signal(SIGXFSZ, SIG_IGN);
-		assert_success(setrlimit(RLIMIT_FSIZE, &rlim));
-		assert_failure(iop_cp(&args));
+		if(setrlimit(RLIMIT_FSIZE, &rlim) != 0)
+		{
+			_Exit(EXIT_FAILURE);
+		}
+
+		/* We expect copy operation to fail. */
+		if(iop_cp(&args) == 0)
+		{
+			_Exit(EXIT_FAILURE);
+		}
 
 		_Exit(EXIT_SUCCESS);
 	}
```
I get some other test failures then, but those tests aren't new, so maybe they fail due to issues with my VM.

**Comment by xaizek** [comment]
And this patch seems to fix kfreebsd-amd64:
```patch
diff --git a/src/int/fuse.c b/src/int/fuse.c
index 4b8abd37f..55f62b26c 100644
--- a/src/int/fuse.c
+++ b/src/int/fuse.c
@@ -669,13 +669,15 @@ run_fuse_command(char cmd[], const cancellation_t *cancellation, int *cancelled)
 static void
 kill_mount_point(const char mount_point[])
 {
-	if(rmdir(mount_point) != 0 && errno == ENOTDIR)
-	{
-		/* unlink() will fail if there is a trailing slash. */
-		char no_slash[strlen(mount_point) + 1];
-		strcpy(no_slash, mount_point);
-		chosp(no_slash);
+	/* rmdir() on some systems (FreeBSD at least) can resolve symbolic links if
+	 * path ends with a slash and unlink() will also fail if there is a trailing
+	 * slash. */
+	char no_slash[strlen(mount_point) + 1];
+	strcpy(no_slash, mount_point);
+	chosp(no_slash);
 
+	if(rmdir(no_slash) != 0 && errno == ENOTDIR)
+	{
 		/* FUSE mounter might replace directory with a symbolic link, account for
 		 * this possibility. */
 		(void)unlink(no_slash);

```

As for armhf, I'm pretty sure it's the same issue you reported for 0.10, which showed up only once. I don't get those failures. They do seem to point at some bug with processing `//` by `:select`, but I couldn't find it by staring at the code, at least not yet.

**Comment by onovy** [comment]
kfreebsd: fixed, thanks!
armhf: agree, it builds now fine (without any change)
hurd: it looks like it's not fixed. I tried it twice and it "hangs" in our building infra. See: https://buildd.debian.org/status/fetch.php?pkg=vifm&arch=hurd-i386&ver=0.10.1-2&stamp=1565539045&raw=0

**Comment by xaizek** [comment]
> hurd: it looks like it's not fixed. I tried it twice and it "hangs" in our building infra.

Strange, the significant functional change there seems to be:
```patch
-		assert_success(setrlimit(RLIMIT_FSIZE, &rlim));
+		if(setrlimit(RLIMIT_FSIZE, &rlim) != 0)
+		{
+			_Exit(EXIT_FAILURE);
+		}
```
But I have no idea how it could produce a hang. Test might work without this change, so you can try to revert this piece. However, I used Jessie version and it might behave differently compared to the builder. Maybe I'll be able to setup a newer version (I think finding ready to use VM image failed, but maybe I can install it from CD image or something; these not officially supported systems are hard to get running).

**Comment by xaizek** [comment]
hurd hangs on `misc/ipc` tests, without the patch they just weren't run. It looks like tests didn't pass there since 0.9.1 which added `--remote-expr`. It's related to opening named pipes and vifm's IPC mechanism seems to not work there. They also didn't hang, but failed prior to 0.10.1.

**Comment by xaizek** [comment]
This is quite outdated. I remember spending a lot of time on this and still having trouble. Should be reopened or reported as a new issue if there are still problems with tests.
